oarepo_doi/services/provider.py

```diff
@@ -163,50 +163,164 @@ def create_and_reserve(self, record, **kwargs):
         json_content = json.loads(content)
         doi_value = json_content["data"]["id"]
         self.add_doi_value(record, record, doi_value)
+        parent_doi = self.get_pid_doi_value(record, parent=True)
 
         if "event" in kwargs:
             pid_status = 'R'
-            parent_doi = self.get_pid_doi_value(record, parent=True)
+
             if parent_doi is None:
 
-                self.register_parent_doi(record, request_metadata, username, password, prefix)
+                parent_doi = self.register_parent_doi(record, request_metadata, username, password, prefix, doi_value)
             elif parent_doi and record.versions.is_latest:
-                self.update_parent_doi(record, request_metadata, username, password)
+                self.update_parent_doi(record, request_metadata, username, password, doi_value)
         else:
             pid_status = 'K'
+        if parent_doi and record.is_published:
+            self.update_relations( parent_doi, record)
+
 
         BaseProvider.create('doi', doi_value, 'rec', record.id, pid_status)
         db.session.commit()
 
-    def register_parent_doi(self, record, request_metadata, username, password, prefix):
-        request_metadata["data"]["attributes"]["prefix"] = str(prefix)
-        request_metadata["data"]["attributes"]["event"] = "publish"
+    def update_relations(self, parent_doi, record):
+
+
+        creds = self.credentials(record)
+        if creds is None:
+            raise ValidationError("No credentials provided.")
+        username, password, _ = creds
+        doi_value = self.get_doi_value(record)
+        url = self.url.rstrip("/") + "/" + doi_value.replace("/", "%2F")
+
+        new_data = requests.get(
+            url=url,
+        )
+
+        if "data" in new_data.json():
+            new_related_identifiers = new_data.json()["data"]["attributes"]["relatedIdentifiers"]
+        else:
+            new_related_identifiers = []
+        if type(parent_doi) is not str:
+            parent_doi = parent_doi.pid_value
+        if not any(item['relatedIdentifier'] == parent_doi for item in new_related_identifiers):
```

Comment:
Please create an `add_relation(identifiers, type, identifier): bool` function and use it to avoid repeated if any(...). Should return if the array was modified.

Use it as:

```
new_version_modified_relations_count += add_relation(new_version_related_identifiers, "IsVersionOf", parent_doi)
```

and save only if the count>0